docs/features/nullable-metadata.md

```diff
@@ -0,0 +1,191 @@
+`NullableContextAttribute` attributes to emit.
+First, `NullableAttribute` attributes are generated at each type reference and type parameter definition by:
+calculating the `byte[]`, skipping empty `byte[]`, and collapsing `byte[]` to single `byte` where possible.
+Then at each level in metadata hierarchy starting at methods:
+The compiler finds the most common single `byte` value across all the `NullableAttribute` attributes at that level
+and any `NullableContextAttribute` attributes on immediate children.
+If there are no single `byte` values, there are no changes.
+Otherwise, a `NullableContext(value)` attribute is created at that level where `value` is most common
+value (preferring `0` over `1` and preferring `1` over `2`), and all `NullableAttribute` and `NullableContextAttribute` attributes with that value are removed.
+That iterative process continues up to the module level.
+If the common value at the module level is a value other than `0` (the default), a module level `NullableContext(value)` attribute is emitted.
+
+Note that an assembly compiled with C#8 where all reference types are oblivious will have no
+`NullableContextAttribute` and no `NullableAttribute` attributes emitted.
+That is equivalent to a legacy assembly.
+
+### Examples
+```C#
+// C# representation of metadata
+[NullableContext(2)]
+class Program
+{
+    string s;                        // string?
+    [Nullable({ 2, 1, 2 }] Dictionary<string, object> d; // Dictionary<string!, object?>?
+    [Nullable(1)] int[] a;           // int[]!
+    int[] b;                         // int[]?
+    [Nullable({ 0, 2 })] object[] c; // object?[]~
+}
+```
+
+## NullableMembersAttribute
+
+`NullableMembersAttribute` can be used to indicate whether nullable attributes were included for `private` or `internal` members.
+
+```C#
+namespace System.Runtime.CompilerServices
+{
+    public enum NullableMembers
+    {
+        Public = 0,   // public and protected only
+        Internal = 1, // public, protected, internal
+        All = 2,
+    }
+    [System.AttributeUsage(AttributeTargets.Module, AllowMultiple = false)]
+    public sealed class NullableMembersAttribute : Attribute
+    {
+        public readonly NullableMembers Members;
+        public NullableMembersAttribute(NullableMembers members)
+        {
+            Members = members;
+        }
+    }
+}
+```
+
+The `NullableMembersAttribute` type is for compiler use only - it is not permitted in source.
+The type declaration is synthesized by the compiler if not already included in the compilation.
+
+The `NullableMembersAttribute` must be emitted if nullable attributes are dropped for
+`private` or `internal` members to allow tools to correctly interpret the nullability of members
+without explicit `NullableAttribute` attributes.
+
+To reduce the size of metadata, the C#8 compiler will not emit attributes for `private` members,
+and the compiler will only emit attributes for `internal` members if the assembly contains
```

Comment:
>assembly [](start = 73, length = 8)

module?

Reply:
`InternalsVisibleToAttribute` is an assembly-level attribute.

---
In reply to: [293554207](https://github.com/dotnet/roslyn/pull/36398#discussion_r293554207) [](ancestors = 293554207)